### PR TITLE
Open pdb files with FileShare.Read and FileShare.Delete

### LINKED ISSUE #38212: PDB files are locked and can't be deleted/renamed once loaded
We've been getting reports of locked pdb files (https://github.com/aspnet/AspNetCore/issues/3269) when deploying ASP.NET Core applications (but not locked dlls) and we assumed that it was application insights locking the pdb. Turns out, the runtime opens pdb files (once somebody accesses something like Exception.StackTrace) it can't be renamed/deleted until the process ends. This interrupts the deployment process on azure app service. It turns out we need to pass the right FileShare permissions when opening the file pdb file.

Here's a simple repro:

```C#
using System;
using System.IO;

namespace ConsoleApp52
{
    class Program
    {
        static void Main(string[] args)
        {
            var pdbFile = Path.ChangeExtension(typeof(Program).Assembly.Location, ".pdb");
            var assemblyFile = typeof(Program).Assembly.Location;

            try
            {
                throw new Exception();
            }
            catch (Exception e)
            {
                // This forces the PDB to be read
                Console.WriteLine(e.StackTrace);
            }

            File.Move(assemblyFile, Path.ChangeExtension(assemblyFile, ".dll.delete"));
            // This line fails
            File.Move(pdbFile, Path.ChangeExtension(assemblyFile, ".pdb.delete"));

            Console.ReadKey();
        }
    }
}
```

I think this fix should be ported down level as well.

### PR DESCRIPTION
**Pull request by davidfowl**
Fixes #38212

PS: I don't know how to test this. I didn't find any existing tests for this class.

### CODE REVIEW

**Comment by tmat** [comment]
LGTM

**Comment by tmat** [comment]
> I don't know how to test this. I didn't find any existing tests for this class.

Couldn't you write a test that does the same as the code snippet in #38212?

**Comment by davidfowl** [comment]
> Couldn't you write a test that does the same as the code snippet in #38212?

I'd be worried about how things like helix deployed tests. Maybe the pdb is always there so I don't need to worry?

**Comment by tmat** [comment]
You can follow the pattern used in System.Reflection.Metadata tests - embed the PDB to the unit test assembly like so:
https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/tests/System.Reflection.Metadata.Tests.csproj#L128-L129

The tests can then extract the PDB from the resource and for example, copy it to a temp directory:
See https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/tests/Resources/TestResources.cs

File: src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs [hunk 1]
```diff
@@ -210,9 +210,15 @@ void IDisposable.Dispose()
                 return null;
             }
 
+            if (string.IsNullOrEmpty(path))
+            {
+                return null;
+            }
```
Comment (stephentoub) [inline src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: Why this change?

Comment (davidfowl) [reply src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: https://dev.azure.com/dnceng/public/_build/results?buildId=211548&view=logs&jobId=77593f2f-9c1d-5a27-f56c-e4d6cda477a7

Comment (davidfowl) [reply src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: Should I just bang it away?

Comment (stephentoub) [reply src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: > Should I just bang it away?

Yup.

Comment (stephentoub) [reply src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: (When we eventually attribute File.Exists, the argument will be attributed as [NotNullWhen(true)], and then the ! won't be necessary.)

**Comment by stephentoub** [comment]
Thanks for fixing it, David.

**Comment by stephentoub** [comment]
Also, I assume you only saw an issue here on Windows, not on Linux, right?

**Comment by davidfowl** [comment]
> If we could add a test, that'd be great.

Will look into it. To reproduce the true path I would want to observe the exception's stack trace or something similar. That would mean triggering an exception that would lead to reading the stack trace from a pdb that I could access on disk. I'm not sure how easy that will be yet, even with the approach @tmat mentioned. I might be overthinking it though.

> Thanks for fixing it, David. Are there other cases similar to this that we need to fix, too, beyond PDBs? Do we have any related issues with .dlls in the runtime, or anything like that?

Haven't looked. I did do a scan for OpenRead when looking at this fix but nothing stuck out as yet. The problematic cases are when the Stream isn't immediately disposed after use.

> Also, I assume you only saw an issue here on Windows, not on Linux, right?

Right, the reported issues were all on windows.

**Comment by davidfowl** [comment]
All failures are unrelated, merging this one now.

File: src/System.Diagnostics.StackTrace/tests/StackTraceSymbolsTests.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public class StackTraceSymbolsTests
+    {
+        [Fact]
+        public void StackTraceSymbolsDoNotLockFile()
+        {
+            var asmPath = typeof(StackTraceSymbolsTests).Assembly.Location;
+            var pdbPath = Path.ChangeExtension(asmPath, ".pdb");
+
+            Assert.True(File.Exists(pdbPath));
```
Comment (tmat) [inline src/System.Diagnostics.StackTrace/tests/StackTraceSymbolsTests.cs]: This assumes the test project builds with non-embedded PDB, which might not be true. CoreFX might be overriding Arcade SDK defaults, but Arcade uses embedded PDBs for local and PR validation builds. See https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/ProjectDefaults.props#L25-L38

Comment (davidfowl) [reply src/System.Diagnostics.StackTrace/tests/StackTraceSymbolsTests.cs]: I felt so clever coming up with this very simple test 😄 .
